[SPARK-28552][SQL] Case-insensitive database URLs in JdbcDialect

teeyog opened the pull request:
## What changes were proposed in this pull request?
This pr proposes to be case insensitive when matching dialects via jdbc url prefix.

When I use jdbc url such as: ```jdbc: MySQL://localhost/db``` to query data through sparksql, the result is wrong, but MySQL supports such url writing.

because sparksql matches MySQLDialect by prefix ```jdbc:mysql```, so ```jdbc: MySQL``` is not matched with the correct dialect. Therefore, it should be case insensitive when identifying the corresponding dialect through jdbc url

https://issues.apache.org/jira/browse/SPARK-28552
## How was this patch tested?
UT.

Commit:
jdbc dialect can match non-lowercase URL prefixes

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

gatorsmile commented:
ok to test

gatorsmile commented:
cc @maropu

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13522/
Test PASSed.

SparkQA commented:
**[Test build #108426 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108426/testReport)** for PR 25287 at commit [`02f5cf9`](https://github.com/apache/spark/commit/02f5cf999bb35b536f48200d42e01f2e119a9b09).

maropu commented:
@teeyog hi, thanks for your first contribution! btw, can you fix this in a more general way? Probably, can we lowercase a prefix in the JdbcDialect side?  I think its a bit troublesome to add the logic for lowercase in each dialect...

Also, plz add tests in `JDBCSuite`.

maropu commented:
And, could you please make the PR description clearer as much as possible to make the other reviewers understood easily... we don't have a rigid format for that, but I'ld like you to clearly describe what you propose in this pr? e.g., "This pr proposes to ..."

If this pr merged, the description will be included in a commit log, so your kind cooperation makes the commit logs more readable.

SparkQA commented:
**[Test build #108426 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108426/testReport)** for PR 25287 at commit [`02f5cf9`](https://github.com/apache/spark/commit/02f5cf999bb35b536f48200d42e01f2e119a9b09).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108426/
Test PASSed.

teeyog commented:
@maropu Thank you very much for your guidance. I will improve this PR description. In addition, you suggest that URL be converted to lowercase only through JdbcDialect without modifying the logic of dialects. I have not thought of a better way to implement it, or to control the incoming url, which is already lowercase, but I don't think that's very good.

Commit:
jdbc dialect matching is not case sensitive

SparkQA commented:
**[Test build #108454 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108454/testReport)** for PR 25287 at commit [`fa6b8a4`](https://github.com/apache/spark/commit/fa6b8a4ab148a11a0209d3a7408648cbaee46e03).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13550/
Test PASSed.

SparkQA commented:
**[Test build #108454 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108454/testReport)** for PR 25287 at commit [`fa6b8a4`](https://github.com/apache/spark/commit/fa6b8a4ab148a11a0209d3a7408648cbaee46e03).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108454/
Test PASSed.

maropu commented:
For example,
```
@DeveloperApi
@Evolving
abstract class JdbcDialect extends Serializable {

  def urlName: String

  /**
   * Check if this dialect instance can handle a certain jdbc url.
   * @param url the jdbc url.
   * @return True if the dialect can be applied on the given jdbc url.
   * @throws NullPointerException if the url is null.
   */
  def canHandle(url : String): Boolean = {
    url.toLowerCase(Locale.ROOT).startsWith(s"jdbc:$urlName")
  }
...
}

private object PostgresDialect extends JdbcDialect {
  override val urlName: String = "postgresql"
...
}
```
?

Commit:
Merge branch 'master' of https://github.com/apache/spark into sql_dialect

Commit:
jdbc dialect matching is not case sensitive

teeyog commented:
@maropu Thank you for your example, I misunderstood what you mean, I thought you told me not to modify the logic of other dialects.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13601/
Test PASSed.

SparkQA commented:
**[Test build #108507 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108507/testReport)** for PR 25287 at commit [`c89c859`](https://github.com/apache/spark/commit/c89c859d90d95dc1b9e3db4f341c861379398035).

SparkQA commented:
**[Test build #108507 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108507/testReport)** for PR 25287 at commit [`c89c859`](https://github.com/apache/spark/commit/c89c859d90d95dc1b9e3db4f341c861379398035).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108507/
Test FAILed.

Commit:
jdbc dialect matching is not case sensitive

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13602/
Test PASSed.

SparkQA commented:
**[Test build #108508 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108508/testReport)** for PR 25287 at commit [`0db9f75`](https://github.com/apache/spark/commit/0db9f7505f62d54cce6613f879c1f5bfa3cab75b).

SparkQA commented:
**[Test build #108508 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108508/testReport)** for PR 25287 at commit [`0db9f75`](https://github.com/apache/spark/commit/0db9f7505f62d54cce6613f879c1f5bfa3cab75b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108508/
Test FAILed.

Commit:
jdbc dialect matching is not case sensitive

Commit:
jdbc dialect matching is not case sensitive

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13604/
Test PASSed.

SparkQA commented:
**[Test build #108510 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108510/testReport)** for PR 25287 at commit [`7fc974e`](https://github.com/apache/spark/commit/7fc974ea3c99b6fbb274ffa7d73b4c7be9ada246).

SparkQA commented:
**[Test build #108510 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108510/testReport)** for PR 25287 at commit [`7fc974e`](https://github.com/apache/spark/commit/7fc974ea3c99b6fbb274ffa7d73b4c7be9ada246).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108510/
Test FAILed.

Commit:
jdbc dialect matching is not case sensitive

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13605/
Test PASSed.

SparkQA commented:
**[Test build #108511 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108511/testReport)** for PR 25287 at commit [`3b7a689`](https://github.com/apache/spark/commit/3b7a68969fef520960f62f8f88c3225af291da0b).

SparkQA commented:
**[Test build #108511 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108511/testReport)** for PR 25287 at commit [`3b7a689`](https://github.com/apache/spark/commit/3b7a68969fef520960f62f8f88c3225af291da0b).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108511/
Test FAILed.

teeyog commented:
@maropu  Hi, can you help me find out what caused this error?
```
[info] Packaging /home/jenkins/workspace/SparkPullRequestBuilder@2/examples/target/scala-2.12/spark-examples_2.12-3.0.0-SNAPSHOT.jar ...
[info] Done packaging.
[error]  * abstract method dbTag()java.lang.String in class org.apache.spark.sql.jdbc.JdbcDialect is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("org.apache.spark.sql.jdbc.JdbcDialect.dbTag")
java.lang.RuntimeException: spark-sql: Binary compatibility check failed!
	at scala.sys.package$.error(package.scala:27)
	at com.typesafe.tools.mima.plugin.SbtMima$.reportModuleErrors(SbtMima.scala:83)
	at com.typesafe.tools.mima.plugin.MimaPlugin$$anonfun$mimaReportSettings$7$$anonfun$apply$2.apply(MimaPlugin.scala:68)
	at com.typesafe.tools.mima.plugin.MimaPlugin$$anonfun$mimaReportSettings$7$$anonfun$apply$2.apply(MimaPlugin.scala:59)
	at scala.collection.immutable.Map$Map1.foreach(Map.scala:109)
	at com.typesafe.tools.mima.plugin.MimaPlugin$$anonfun$mimaReportSettings$7.apply(MimaPlugin.scala:59)
	at com.typesafe.tools.mima.plugin.MimaPlugin$$anonfun$mimaReportSettings$7.apply(MimaPlugin.scala:44)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[error] (sql/*:mimaReportBinaryIssues) spark-sql: Binary compatibility check failed!
[error] Total time: 32 s, completed Aug 1, 2019 1:40:55 AM
[error] running /home/jenkins/workspace/SparkPullRequestBuilder@2/dev/mima -Phadoop-2.7 -Pkubernetes -Phive-thriftserver -Phadoop-cloud -Pkinesis-asl -Pyarn -Pspark-ganglia-lgpl -Phive -Pmesos ; received return code 1
```

maropu commented:
Can you update `project/MimaExcludes.scala`, too?

Commit:
jdbc dialect matching is not case sensitive

Commit:
Merge branch 'master' of https://github.com/apache/spark into sql_dialect

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13631/
Test PASSed.

SparkQA commented:
**[Test build #108537 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108537/testReport)** for PR 25287 at commit [`530741c`](https://github.com/apache/spark/commit/530741c9b857fd20b43ef1d8c9108ddf3e8af7bd).

SparkQA commented:
**[Test build #108537 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/108537/testReport)** for PR 25287 at commit [`530741c`](https://github.com/apache/spark/commit/530741c9b857fd20b43ef1d8c9108ddf3e8af7bd).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/108537/
Test FAILed.

srowen commented:
`jdbc: MySQL:/...` isn't a valid URI, is it?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

maropu commented:
ping @teeyog

Commit:
Merge branch 'master' of https://github.com/apache/spark into sql_dialect

SparkQA commented:
**[Test build #111867 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111867/testReport)** for PR 25287 at commit [`5913568`](https://github.com/apache/spark/commit/591356813906ab155282798da9221874629db7be).

SparkQA commented:
**[Test build #111867 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111867/testReport)** for PR 25287 at commit [`5913568`](https://github.com/apache/spark/commit/591356813906ab155282798da9221874629db7be).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111867/
Test FAILed.

teeyog commented:
@maropu 
hi, I reported this error when test build. I don't know how to solve it. Can you help me? Thank you.
You can also go to see Console Output.
```
[error] (spark/javaunidoc:doc) javadoc returned nonzero exit code
[error] Total time: 98 s, completed Oct 7, 2019 8:24:15 PM
[error] running /home/jenkins/workspace/SparkPullRequestBuilder/build/sbt -Phadoop-2.7 -Pkubernetes -Phive-thriftserver -Phadoop-cloud -Pkinesis-asl -Pyarn -Pspark-ganglia-lgpl -Phive -Pmesos unidoc ; received return code 1
```

maropu commented:
Can you check the jenkins log?
```
[info] Generating /home/jenkins/workspace/SparkPullRequestBuilder/target/javaunidoc/org/apache/spark/sql/jdbc/AggregatedDialect.html...
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/core/target/java/org/apache/spark/sql/jdbc/JdbcDialect.java:44: error: invalid entity &#xff0c;
[error]    * The tag of the database, it must be lowercase&#xff0c;the subsequent will match
[error]                                                   ^
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/core/target/java/org/apache/spark/sql/jdbc/JdbcDialect.java:47: error: bad use of '>'
[error]    * mysql  url:  jdbc:mysql://localhost/db  --> mysql
[error]                                                ^
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/core/target/java/org/apache/spark/sql/jdbc/JdbcDialect.java:48: error: bad use of '>'
[error]    * oracle url:  jdbc:oracle://localhost/db --> oracle
[error]                                                ^
[error] /home/jenkins/workspace/SparkPullRequestBuilder/sql/core/target/java/org/apache/spark/sql/jdbc/JdbcDialect.java:49: error: bad use of '>'
[error]    * db2    url:  jdbc:db2://localhost/db    --> db2
[error]                                                ^
[info] Generating 
```

Commit:
jdbc dialect matching is not case sensitive

SparkQA commented:
**[Test build #112699 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112699/testReport)** for PR 25287 at commit [`93027ca`](https://github.com/apache/spark/commit/93027caf246d7ac613bf0db0d7200080b16c4e03).

SparkQA commented:
**[Test build #112699 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112699/testReport)** for PR 25287 at commit [`93027ca`](https://github.com/apache/spark/commit/93027caf246d7ac613bf0db0d7200080b16c4e03).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112699/
Test FAILed.

Commit:
jdbc dialect matching is not case sensitive

SparkQA commented:
**[Test build #112701 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112701/testReport)** for PR 25287 at commit [`404bb34`](https://github.com/apache/spark/commit/404bb34e9a3fa8f432bce62823dcde07bdbbdb2f).

SparkQA commented:
**[Test build #112701 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112701/testReport)** for PR 25287 at commit [`404bb34`](https://github.com/apache/spark/commit/404bb34e9a3fa8f432bce62823dcde07bdbbdb2f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112701/
Test FAILed.

Commit:
jdbc dialect matching is not case sensitive

SparkQA commented:
**[Test build #112758 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112758/testReport)** for PR 25287 at commit [`c27dfff`](https://github.com/apache/spark/commit/c27dfff6161f5a99d3743ad486268b92fddba45a).

SparkQA commented:
**[Test build #112758 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112758/testReport)** for PR 25287 at commit [`c27dfff`](https://github.com/apache/spark/commit/c27dfff6161f5a99d3743ad486268b92fddba45a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112758/
Test PASSed.

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala:
```diff
@@ -56,13 +57,26 @@ case class JdbcType(databaseTypeDefinition : String, jdbcNullType : Int)
+   * db2    url:  jdbc:db2://localhost/db    dbTag: db2
+   * ......
+   */
+  def dbTag: String
```
Hm, the problem here is that this will break any custom dialects as it adds a new required method to implement. I am not sure it's worth it, even though we can possibly introduce such breaking changes. Part of the reason I don't think it's worth it is just that a non-lowercase string is pretty nonstandard.

I would either push the lowercase down into each implementation, or provide a default implementation that returns nothing. That at least covers existing implementations which will already override `canHandle`.

teeyog replied:
So how do you look like this, the canHandle in each implementation is cover to lowercase
```
private case object MySQLDialect extends JdbcDialect {

  override def canHandle(url : String): Boolean =
      url.toLowerCase(Locale.ROOT).startsWith("jdbc:mysql")
```

teeyog replied:
Not lowercase is not standard, but the uppercase url can read data correctly from mysql, I think it needs to be compatible here.

srowen replied:
Yes, at best that is what I'd do here. It's not worth breaking the developer API even in a minor way.

maropu replied:
Yea, that new one looks nice to me, too.

jiangxb1987 committed:
Prepare Spark release v3.0.0-preview-rc2

### What changes were proposed in this pull request?

To push the built jars to maven release repository, we need to remove the 'SNAPSHOT' tag from the version name.

Made the following changes in this PR:
* Update all the `3.0.0-SNAPSHOT` version name to `3.0.0-preview`
* Update the sparkR version number check logic to allow jvm version like `3.0.0-preview`

**Please note those changes were generated by the release script in the past, but this time since we manually add tags on master branch, we need to manually apply those changes too.**

We shall revert the changes after 3.0.0-preview release passed.

### Why are the changes needed?

To make the maven release repository to accept the built jars.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

N/A

Commit:
jdbc dialect matching is not case sensitive

Commit:
jdbc dialect matching is not case sensitive

SparkQA commented:
**[Test build #112877 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112877/testReport)** for PR 25287 at commit [`5c495aa`](https://github.com/apache/spark/commit/5c495aa6a2690a439396c4b29d21a5e364077406).

SparkQA commented:
**[Test build #112877 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112877/testReport)** for PR 25287 at commit [`5c495aa`](https://github.com/apache/spark/commit/5c495aa6a2690a439396c4b29d21a5e364077406).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112877/
Test FAILed.

Commit:
dialect

Commit:
Merge branch 'sql_dialect' of https://github.com/teeyog/spark into sql_dialect

Commit:
jdbc dialect matching is not case sensitive

SparkQA commented:
**[Test build #112878 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112878/testReport)** for PR 25287 at commit [`76e80e7`](https://github.com/apache/spark/commit/76e80e79099f1c6bc9637cf1e306ee8bf83cc8ac).

SparkQA commented:
**[Test build #112878 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112878/testReport)** for PR 25287 at commit [`76e80e7`](https://github.com/apache/spark/commit/76e80e79099f1c6bc9637cf1e306ee8bf83cc8ac).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112878/
Test FAILed.

Commit:
fix

SparkQA commented:
**[Test build #112887 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112887/testReport)** for PR 25287 at commit [`626d8a4`](https://github.com/apache/spark/commit/626d8a44ae2fb71187b6e72341fc5432eb5a9fbb).

SparkQA commented:
**[Test build #112887 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112887/testReport)** for PR 25287 at commit [`626d8a4`](https://github.com/apache/spark/commit/626d8a44ae2fb71187b6e72341fc5432eb5a9fbb).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112887/
Test FAILed.

yaooqinn committed:
[SPARK-29653][SQL] Fix MICROS_PER_MONTH in IntervalUtils

### What changes were proposed in this pull request?

MICROS_PER_MONTH = DAYS_PER_MONTH * MICROS_PER_DAY

### Why are the changes needed?

fix bug

### Does this PR introduce any user-facing change?

no
### How was this patch tested?

add ut

Closes #26321 from yaooqinn/SPARK-29653.

Authored-by: Kent Yao <yaooqinn@hotmail.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

maropu committed:
[SPARK-29120][SQL][TESTS] Port create_view.sql

### What changes were proposed in this pull request?

This PR ports create_view.sql from PostgreSQL regression tests https://github.com/postgres/postgres/blob/REL_12_STABLE/src/test/regress/sql/create_view.sql

The expected results can be found in the link: https://github.com/postgres/postgres/blob/REL_12_STABLE/src/test/regress/expected/create_view.out

### Why are the changes needed?

To check behaviour differences between Spark and PostgreSQL

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

Pass the Jenkins. And, Comparison with PgSQL results

Closes #26290 from maropu/SPARK-29120.

Authored-by: Takeshi Yamamuro <yamamuro@apache.org>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

dongjoon-hyun committed:
[SPARK-29668][DOCS] Deprecate Python 3 prior to version 3.6

### What changes were proposed in this pull request?

This PR aims to deprecate `Python 3.4 ~ 3.5`, which is prior to version 3.6 additionally.

### Why are the changes needed?

Since `Python 3.8` is already out, we will focus on to support Python 3.6/3.7/3.8.

### Does this PR introduce any user-facing change?

Yes. It's highly recommended to use Python 3.6/3.7. We will verify Python 3.8 before Apache Spark 3.0.0 release.

### How was this patch tested?

NA (This is a doc-only change).

Closes #26326 from dongjoon-hyun/SPARK-29668.

Authored-by: Dongjoon Hyun <dhyun@apple.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

jiangxb1987 committed:
[SPARK-29646][BUILD] Allow pyspark version name format `${versionNumber}-preview` in release script

### What changes were proposed in this pull request?

Update `release-build.sh`, to allow pyspark version name format `${versionNumber}-preview`, otherwise the release script won't generate pyspark release tarballs.

### How was this patch tested?

Tested locally.

Closes #26306 from jiangxb1987/buildPython.

Authored-by: Xingbo Jiang <xingbo.jiang@databricks.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

jiangxb1987 committed:
[SPARK-29666][BUILD] Fix the publish release failure under dry-run mode

### What changes were proposed in this pull request?

`release-build.sh` fail to publish release under dry run mode with the following error message:
```
/opt/spark-rm/release-build.sh: line 429: pushd: spark-repo-g4MBm/org/apache/spark: No such file or directory
```

We need to at least run the `mvn clean install` command once to create the `$tmp_repo` path, but now those steps are all skipped under dry-run mode. This PR fixes the issue.

### How was this patch tested?

Tested locally.

Closes #26329 from jiangxb1987/dryrun.

Authored-by: Xingbo Jiang <xingbo.jiang@databricks.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

jiangxb1987 committed:
Revert "Prepare Spark release v3.0.0-preview-rc2"

This reverts commit 007c873ae34f58651481ccba30e8e2ba38a692c4.

rdblue committed:
[SPARK-29277][SQL] Add early DSv2 filter and projection pushdown

### What changes were proposed in this pull request?

This adds a new rule, `V2ScanRelationPushDown`, to push filters and projections in to a new `DataSourceV2ScanRelation` in the optimizer. That scan is then used when converting to a physical scan node. The new relation correctly reports stats based on the scan.

To run scan pushdown before rules where stats are used, this adds a new optimizer override, `earlyScanPushDownRules` and a batch for early pushdown in the optimizer, before cost-based join reordering. The other early pushdown rule, `PruneFileSourcePartitions`, is moved into the early pushdown rule set.

This also moves pushdown helper methods from `DataSourceV2Strategy` into a util class.

### Why are the changes needed?

This is needed for DSv2 sources to supply stats for cost-based rules in the optimizer.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

This updates the implementation of stats from `DataSourceV2Relation` so tests will fail if stats are accessed before early pushdown for v2 relations.

Closes #25955 from rdblue/move-v2-pushdown.

Authored-by: Ryan Blue <blue@apache.org>
Signed-off-by: Ryan Blue <blue@apache.org>

MaxGekk committed:
[SPARK-29651][SQL] Fix parsing of interval seconds fraction

### What changes were proposed in this pull request?
In the PR, I propose to extract parsing of the seconds interval units to the private method `parseNanos` in `IntervalUtils` and modify the code to correctly parse the fractional part of the seconds unit of intervals in the cases:
- When the fractional part has less than 9 digits
- The seconds unit is negative

### Why are the changes needed?
The changes are needed to fix the issues:
```sql
spark-sql> select interval '10.123456 seconds';
interval 10 seconds 123 microseconds
```
The correct result must be `interval 10 seconds 123 milliseconds 456 microseconds`
```sql
spark-sql> select interval '-10.123456789 seconds';
interval -9 seconds -876 milliseconds -544 microseconds
```
but the whole interval should be negated, and the result must be `interval -10 seconds -123 milliseconds -456 microseconds`, taking into account the truncation to microseconds.

### Does this PR introduce any user-facing change?
Yes. After changes:
```sql
spark-sql> select interval '10.123456 seconds';
interval 10 seconds 123 milliseconds 456 microseconds
spark-sql> select interval '-10.123456789 seconds';
interval -10 seconds -123 milliseconds -456 microseconds
```

### How was this patch tested?
By existing and new tests in `ExpressionParserSuite`.

Closes #26313 from MaxGekk/fix-interval-nanos-parsing.

Authored-by: Maxim Gekk <max.gekk@gmail.com>
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

Commit:
Merge remote-tracking branch 'upstream/master' into sql_dialect

d80tb7 committed:
[SPARK-29126][PYSPARK][DOC] Pandas Cogroup udf usage guide

This PR adds some extra documentation for the new Cogrouped map Pandas udfs.  Specifically:

- Updated the usage guide for the new `COGROUPED_MAP` Pandas udfs added in https://github.com/apache/spark/pull/24981
- Updated the docstring for pandas_udf to include the COGROUPED_MAP type as suggested by HyukjinKwon in https://github.com/apache/spark/pull/25939

Closes #26110 from d80tb7/SPARK-29126-cogroup-udf-usage-guide.

Authored-by: Chris Martin <chris@cmartinit.co.uk>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

Commit:
Merge remote-tracking branch 'upstream/master'

SparkQA commented:
**[Test build #112983 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112983/testReport)** for PR 25287 at commit [`28af8f9`](https://github.com/apache/spark/commit/28af8f9571e64bd974cec5c2c299520c20267824).

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -1649,4 +1649,50 @@ class JDBCSuite extends QueryTest
+      matchingDialects.length match {
+        case 0 => NoopDialect
+        case 1 => matchingDialects.head
+        case _ => new AggregatedDialect(matchingDialects)
```
nit: we need the entries for `0` and `_`? I think the second one is only meaningful for this test?

maropu reviewed (approved):
LGTM @srowen @HyukjinKwon

SparkQA commented:
**[Test build #112983 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112983/testReport)** for PR 25287 at commit [`28af8f9`](https://github.com/apache/spark/commit/28af8f9571e64bd974cec5c2c299520c20267824).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112983/
Test PASSed.

Commit:
Merge branch 'master' of https://github.com/apache/spark

Commit:
Merge branch 'master' of https://github.com/apache/spark

teeyog committed:
Merge branch 'master' into sql_dialect

SparkQA commented:
**[Test build #113075 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113075/testReport)** for PR 25287 at commit [`a5148f0`](https://github.com/apache/spark/commit/a5148f0fd967e3157d1cfdb856b3f89fd793f24b).

SparkQA commented:
**[Test build #113075 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113075/testReport)** for PR 25287 at commit [`a5148f0`](https://github.com/apache/spark/commit/a5148f0fd967e3157d1cfdb856b3f89fd793f24b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait HasRelativeError extends Params `
  * `class _ImputerParams(HasInputCol, HasInputCols, HasOutputCol, HasOutputCols, HasRelativeError):`
  * `class _RobustScalerParams(HasInputCol, HasOutputCol, HasRelativeError):`
  * `class HasRelativeError(Params):`
  * `case class AlterTableRenamePartitionStatement(`
  * `case class ShowColumnsStatement(`
  * `case class DataSourceV2ScanRelation(`
  * `case class OptimizeLocalShuffleReader(conf: SQLConf) extends Rule[SparkPlan] `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113075/
Test FAILed.

Commit:
fix conflicts

Commit:
fix conflict

SparkQA commented:
**[Test build #113078 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113078/testReport)** for PR 25287 at commit [`3e58b81`](https://github.com/apache/spark/commit/3e58b81569ba3a6de7f39642effc45a67e73bfb8).

SparkQA commented:
**[Test build #113078 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113078/testReport)** for PR 25287 at commit [`3e58b81`](https://github.com/apache/spark/commit/3e58b81569ba3a6de7f39642effc45a67e73bfb8).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113078/
Test FAILed.

Commit:
fix conflict

SparkQA commented:
**[Test build #113079 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113079/testReport)** for PR 25287 at commit [`ac70da4`](https://github.com/apache/spark/commit/ac70da4f494624368aa1570100e2a00ec0e1c308).

SparkQA commented:
**[Test build #113079 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113079/testReport)** for PR 25287 at commit [`ac70da4`](https://github.com/apache/spark/commit/ac70da4f494624368aa1570100e2a00ec0e1c308).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113079/
Test PASSed.

srowen reviewed (approved):
I'm OK with it. I looked it up and technically URI schemes are meant to be case insensitive, so that's a good argument for this change.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -58,7 +58,7 @@ class JDBCSuite extends QueryTest
 
   val testH2DialectTinyInt = new JdbcDialect {
-    override def canHandle(url: String): Boolean = url.startsWith("jdbc:h2")
+    override def canHandle(url: String) : Boolean = url.startsWith("jdbc:h2")
```
nit: revert this.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -1662,4 +1662,50 @@ class JDBCSuite extends QueryTest
       "`NONE`, `READ_UNCOMMITTED`, `READ_COMMITTED`, `REPEATABLE_READ` or `SERIALIZABLE`."))
   }
+
+  test("SPARK-28552: Check whether a dialect instance can be applied on the given jdbc url") {
```
You can simplify the tests below;
```
  test("SPARK-28552: Case-insensitive database URLs in JdbcDialect") {
    assert(JdbcDialects.get("jdbc:mysql://localhost/db") === MySQLDialect)
    assert(JdbcDialects.get("jdbc:MySQL://localhost/db") === MySQLDialect)
    assert(JdbcDialects.get("jdbc:postgresql://localhost/db") === PostgresDialect)
    assert(JdbcDialects.get("jdbc:postGresql://localhost/db") === PostgresDialect)
    assert(JdbcDialects.get("jdbc:db2://localhost/db") === DB2Dialect)
    assert(JdbcDialects.get("jdbc:DB2://localhost/db") === DB2Dialect)
    assert(JdbcDialects.get("jdbc:sqlserver://localhost/db") === MsSqlServerDialect)
    assert(JdbcDialects.get("jdbc:sqlServer://localhost/db") === MsSqlServerDialect)
    assert(JdbcDialects.get("jdbc:derby://localhost/db") === DerbyDialect)
    assert(JdbcDialects.get("jdbc:derBy://localhost/db") === DerbyDialect)
    assert(JdbcDialects.get("jdbc:oracle://localhost/db") === OracleDialect)
    assert(JdbcDialects.get("jdbc:Oracle://localhost/db") === OracleDialect)
    assert(JdbcDialects.get("jdbc:teradata://localhost/db") === TeradataDialect)
    assert(JdbcDialects.get("jdbc:Teradata://localhost/db") === TeradataDialect)
  }
```

maropu reviewed (changes requested):
Could you check my comments before merging?

teeyog commented:
> Could you check my comments before merging?

ok,thanks

Commit:
Merge branch 'master' of https://github.com/apache/spark

Commit:
update

Commit:
update

Commit:
update

Commit:
update

Commit:
update

SparkQA commented:
**[Test build #113195 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113195/testReport)** for PR 25287 at commit [`3e1585b`](https://github.com/apache/spark/commit/3e1585b8f3ceeaf1488cc3af30342892a270d216).

SparkQA commented:
**[Test build #113195 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113195/testReport)** for PR 25287 at commit [`3e1585b`](https://github.com/apache/spark/commit/3e1585b8f3ceeaf1488cc3af30342892a270d216).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113195/
Test FAILed.

Commit:
Merge branch 'master' of https://github.com/apache/spark

Commit:
Merge branch 'sql_dialect' of https://github.com/teeyog/spark into sql_dialect

SparkQA commented:
**[Test build #113196 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113196/testReport)** for PR 25287 at commit [`1883847`](https://github.com/apache/spark/commit/188384741876e88b6d79716283c5007fae948f83).

SparkQA commented:
**[Test build #113196 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113196/testReport)** for PR 25287 at commit [`1883847`](https://github.com/apache/spark/commit/188384741876e88b6d79716283c5007fae948f83).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/113196/
Test PASSed.

maropu commented:
Thanks for your first contribution, @teeyog! Merged to master.

maropu commented:
FYI: Added @teeyog in the Spark contributor list.

teeyog commented:
Thanks! @maropu  @srowen
